### DISCUSSION TITLE
Lifecycle of a ProviderRef

### DISCUSSION THREAD

**Discussion by derolf** [rrousselGit/riverpod#2537] [Q&A] [answered]
In the docs, I am reading that Providers get disposed when invalidated, e.g. all handlers registered with ProviderRef.onDispose get called.

What about the ProviderRef? Will it be always the same if the Provider is rebuild at the same frame like it was disposed?

**Comment by rrousselGit** [comment]
I do not understand the question. I don't see what you mean by ProviderRef in this context.

**Reply by derolf** [reply]
I mean the ProviderRef that is passed as ref to the Provider.

**Reply by rrousselGit** [reply]
But what about it?

**Reply by derolf** [reply]
Just a dumb example to clarify what I mean:

```
final refProvider = Provider((ProviderRef ref) => ref);

// here they should be different
void differentIfDead() {
  container = ProviderContainer();

  final ref1 = container.read(refProvider);

  container.pump();

  final ref2 = container.read(refProvider);

  assert(ref1 != ref2);
}

// here they should be the same
void sameIfListened() {
  container = ProviderContainer();

  // keep alive
  container.listen(refProvider, (_, __) {});

  final ref1 = container.read(refProvider);

  // fore rebuild
  container.invalidate(refProvider);

  container.pump();

  final ref2 = container.read(refProvider);

  assert(ref1 == ref2);
}
```

**Reply by rrousselGit** [reply, marked as answer]
You'll have to await "pump" and make your provider autoDispose for your differentIfDead() if you want them to be different.

Non autoDispose providers never change ref.

**Reply by derolf** [reply]
Yes, need to await the pump, but I think the example made clear what I mean. So basically a ProviderRef stays the same as long as the underlying element exists.

**Reply by rrousselGit** [reply]
Yes. The ref _Is_ the Element after all.

**Reply by derolf** [reply]
Ah, like BuildContext being the Element.

**Reply by rrousselGit** [reply]
But technically there's no guarantee on when a new ref is created.
I've considered making a new instance sometimes.

**Reply by derolf** [reply]
Making a new Ref each time might be good so that people don't intentionally leak Refs outside the build function (like in my example).

**Reply by rrousselGit** [reply]
It's not a huge deal anyway.

I was considering it because then ref.mounted could point to "was the provider invalidated" instead of "did the Element get destroyed".

Although it wouldn't work for Notifier, so it doesn't solve the mounted issue for those.

**Reply by derolf** [reply]
The reason I was opening this thread is that I leaked dead refs through async futures resulting in crazy error messages (like broken dependency overrides although I did override nothing).

**Reply by rrousselGit** [reply]
Oh, I see. Returning refs like this is a bit hacky anyway
